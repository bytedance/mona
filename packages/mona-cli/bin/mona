#! /usr/bin/env node

const mona = require('../dist').default;
mona();
