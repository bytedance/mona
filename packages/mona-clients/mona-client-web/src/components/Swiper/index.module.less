.swiper {
  height: 150px;
  user-select: none;
  -webkit-user-select: none;
  margin-left: auto;
  margin-right: auto;
  overflow: hidden;
  position: relative;
  padding: 0;
  z-index: 1;

  & .wrapper {
    position: relative;
    width: 100%;
    height: 100%;
    display: flex;
    transition-property: transform;
    transition-timing-function: ease-in-out;
  }

  .dots {
    position: absolute;
    font-size: 0;

    &.horizontal {
      left: 50%;
      bottom: 10px;
      text-align: center;
      white-space: nowrap;
      transform: translate(-50%, 0);
    }
  }
  
  .dot {
    display: inline-block;
    margin-right: 8px;
    height: 8px;
    width: 8px;
    border-radius: 50%;
    cursor: pointer;
    transition-property: background-color;
    transition-timing-function: ease;
    background: rgba(0, 0, 0, 0.3);
    border-radius: 50%;

    &:last-child {
      margin-right: 0;
    }
  }
}