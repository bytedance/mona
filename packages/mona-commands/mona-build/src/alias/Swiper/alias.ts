import { bubbleEventsAlias, basePropAlias, propAliasMap } from '../prop';
import { SwiperAlias as Alias } from '../type';

const alias: Alias = {
  ...basePropAlias,
  ...bubbleEventsAlias,
  indicatorDots: propAliasMap.indicatorDots,
  indicatorColor: propAliasMap.indicatorColor,
  indicatorActiveColor: propAliasMap.indicatorActiveColor,
  autoplay: propAliasMap.autoplay,
  current: propAliasMap.current,
  currentItemId: propAliasMap.currentItemId,
  interval: propAliasMap.interval,
  previousMargin: propAliasMap.previousMargin,
  nextMargin: propAliasMap.nextMargin,
  displayMultipleItems: propAliasMap.displayMultipleItems,
  duration: propAliasMap.duration,
  circular: propAliasMap.circular,
  vertical: propAliasMap.vertical,
  onChange: propAliasMap.onChange,
  onAnimationFinish: propAliasMap.onAnimationFinish,
  onTransition: propAliasMap.onTransition,
};

export default alias;
