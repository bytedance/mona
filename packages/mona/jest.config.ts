import baseConfig from '../../jestconfig.base';

module.exports = baseConfig;
