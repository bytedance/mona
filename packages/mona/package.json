{
  "name": "@bytedance/mona",
  "version": "0.2.0-alpha.9",
  "description": "Merchant app's developing and building tools",
  "main": "./dist/index",
  "scripts": {
    "build": "yarn clean && ttsc -b",
    "start": "ttsc -b -w",
    "clean": "rm -rf dist",
    "test": "jest"
  },
  "files": [
    "dist"
  ],
  "author": "zhangxiaowei.chris@bytedance.com",
  "license": "MIT",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "devDependencies": {
    "copy-webpack-plugin": "^10.2.0",
    "react": "^17.0.2",
    "webpack": "^5.58.2",
    "webpack-chain": "^6.5.1"
  },
  "peerDependencies": {
    "react": "^17.0.2"
  },
  "gitHead": "cc5aa5283af595fd7d8a7b4921f54e2797504edd"
}
