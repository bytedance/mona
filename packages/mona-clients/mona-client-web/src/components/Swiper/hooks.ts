import { useEffect, useRef, useState } from 'react';
import { SwiperChangeEvent } from '@bytedance/mona';
import { formatTouchEvent } from '../utils';

let ANIMATION_READY = false;

interface SetTransformParams {
  ele: HTMLElement | null;
  current: number;
  vertical: boolean;
  circular: boolean;
  duration: number;
  noAnimation: boolean;
}

function setTransform({ ele, current, vertical, circular, duration, noAnimation }: SetTransformParams) {
  if (ele) {
    const width = vertical ? ele.clientHeight : ele.clientWidth;
    let realCurrent = current;
    if (circular) {
      realCurrent = current + 1;
    }

    const distance = -width * realCurrent;
    const targetTransform = `${vertical ? 0 : distance}px, ${!vertical ? 0 : distance}px, 0px`;

    ele.style.transitionDuration = `${noAnimation ? 0 : ANIMATION_READY ? duration : 0}ms`;
    ele.style.transform = `translate3d(${targetTransform}`;

    ANIMATION_READY = true;

    if (!noAnimation) {
      setTimeout(() => {
        ele.style.transitionDuration = '0ms';
      }, duration);
    }
  }
}

const tick = 100;

interface UseCurrentParams {
  initCurrent: number;
  total: number;
  duration: number;
  circular: boolean;
}

function useCurrent({ initCurrent, total, duration, circular }: UseCurrentParams) {
  const [current, setCurrent] = useState(initCurrent);
  const lockRef = useRef(false);
  const noAnimationRef = useRef(true);
  const min = circular ? -1 : 0;
  const max = total ? (circular ? total : total - 1) : 0;

  const safelySetCurrent = (current: number) => {
    if (lockRef.current) {
      return;
    }
    noAnimationRef.current = false;
    lockRef.current = true;
    const target = current < min ? min : current > max ? max : current;
    setCurrent(target);

    setTimeout(() => {
      if (circular) {
        if (target === min) {
          noAnimationRef.current = true;
          setCurrent(total - 1);
        } else if (target === max) {
          noAnimationRef.current = true;
          setCurrent(0);
        }
      }
      lockRef.current = false;
    }, duration + tick);
  };

  useEffect(() => {
    safelySetCurrent(initCurrent);
  }, [initCurrent]);

  return { current, setCurrent: safelySetCurrent, noAnimation: noAnimationRef.current };
}

// add two fake childEle in the tail and head of the container
function preHandleEle(ele: HTMLDivElement | null, circular: boolean) {
  if (!ele || ele.children.length === 0) {
    return;
  }

  if (circular) {
    const firstChild = ele.children[0];
    const lastChild = ele.children[ele.children.length - 1];
    const copyedFirstChild = firstChild.cloneNode(true);
    const coypedLastChild = lastChild.cloneNode(true);

    ele.appendChild(copyedFirstChild);
    ele.insertBefore(coypedLastChild, firstChild);

    return () => {
      (copyedFirstChild as Element).remove();
      (coypedLastChild as Element).remove();
    };
  }
  return () => {};
}

interface UseSlideParams {
  current: number;
  total: number;
  vertical: boolean;
  duration: number;
  circular: boolean;
  onChange?: (event: SwiperChangeEvent) => void;
}

export function useSlide({
  current: initCurrent,
  total,
  vertical = false,
  duration,
  circular,
  onChange,
}: UseSlideParams) {
  const ref = useRef<HTMLDivElement | null>(null);
  const { current, setCurrent, noAnimation } = useCurrent({ initCurrent, total, duration, circular });

  useEffect(() => preHandleEle(ref.current, circular), [ref.current, circular]);

  useEffect(() => {
    setTransform({ ele: ref.current, current, vertical, circular, duration, noAnimation });
  }, [ref.current, current, vertical, circular, duration, noAnimation]);

  const activeIndex = current < 0 ? total - 1 : current >= total ? 0 : current;

  const setActiveIndex = (index: number, byAuto?: boolean, e?: React.TouchEvent) => {
    if (typeof onChange === 'function') {
      const event: SwiperChangeEvent = e ?
        {
          ...formatTouchEvent({ event: e, type: 'change' }),
          detail: {
            current: activeIndex,
            source: byAuto ? 'autoplay' : 'touch',
          },
        } :
        {
          touches: [],
          changedTouches: [],
          type: 'change',
          target: { id: '', tagName: '', dataset: {} },
          currentTarget: { id: '', tagName: '', dataset: {} },
          timeStamp: 0,
          detail: {
            current: activeIndex,
            source: byAuto ? 'autoplay' : 'touch',
          },
        };
      onChange(event);
    }
    setCurrent(index >= total ? 0 : index < 0 ? total - 1 : index);
  };

  const next = (byAuto?: boolean, e?: React.TouchEvent) => {
    setActiveIndex(current + 1, byAuto, e);
  };

  const prev = (byAuto?: boolean, e?: React.TouchEvent) => {
    setActiveIndex(current - 1, byAuto, e);
  };

  return {
    next,
    prev,
    wrapperRef: ref,
    activeIndex,
  };
}

type HandleFunc = ((e: React.TouchEvent) => void) | null;
interface UseTouchParams {
  left: HandleFunc;
  right: HandleFunc;
  top: HandleFunc;
  bottom: HandleFunc;
  duration: number;
}

const validTouchDistance = 20;
export function useTouch({ left, right, top, bottom, duration }: UseTouchParams) {
  const startRef = useRef<{ x: number; y: number }>({ x: 0, y: 0 });
  const lockRef = useRef(false);
  const handleTouchStart = (e: React.TouchEvent) => {
    const { changedTouches } = e;
    startRef.current = {
      x: changedTouches[0].pageX,
      y: changedTouches[0].pageY,
    };
  };

  const handleTouchMove = (e: React.TouchEvent) => {
    if (lockRef.current) {
      return;
    }
    const { changedTouches } = e;
    const moveX = changedTouches[0].pageX;
    const moveY = changedTouches[0].pageY;
    const { x, y } = startRef.current;

    const dx = moveX - x;
    const dy = moveY - y;

    if (Math.abs(dx) >= Math.abs(dy) && Math.abs(dx) > validTouchDistance) {
      lockRef.current = true;

      // horizontal move
      if (dx > 0) {
        // l2r
        right && right(e);
      } else {
        // r2l
        left && left(e);
      }
      setTimeout(() => {
        lockRef.current = false;
      }, duration + tick);
    } else if (Math.abs(dx) < Math.abs(dy) && Math.abs(dy) > validTouchDistance) {
      lockRef.current = true;
      // vertical move
      if (dy > 0) {
        bottom && bottom(e);
      } else {
        top && top(e);
      }
      setTimeout(() => {
        lockRef.current = false;
      }, duration + tick);
    }
  };

  return { handleTouchStart, handleTouchMove };
}
