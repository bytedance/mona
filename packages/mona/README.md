# @bytedance/mona
对外暴露提供构建时API
## 配置创建
### createProjectConfig
创建项目配置，`mona.config.ts`中使用
#### 例子
```js
// mona.config.ts
import { createProjectConfig } from '@bytedance/mona';
export default createProjectConfig({
  projectName: 'demo',
  input: './src/app.tsx',
  output: 'dist'
})
```

#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| projectName  | 应用/插件名称 | 是 | string | - |
| input  | 入口文件路径 | 否 | string | - |
| output  | 打包后的文件目录路径 | 否 | string | - |
| raw  | 定义函数来修改webpack配置 | 否 | (options: any) => any | - |
| dev  | 用来修改本地服务配置 | 否 | string | { port: 9999 } |

### createAppConfig
创建应用配置，`app.config.ts`中使用
#### 例子
```js
// app.config.ts
import { createAppConfig } from '@bytedance/mona';
export default createAppConfig({
  pages: [
    'pages/Home/index',
    'pages/Info/index',
    'pages/List/index'
  ]
})
```
#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| pages  | 页面路径 | 是 | string[] | - |

默认`pages`的第一项为首页
### createPageConfig
创建页面配置，需在每个页面目录的`page.config.ts`中使用

#### 例子
```js
// page.config.ts
import { createPageConfig } from '@bytedance/mona';
export default createPageConfig({
  navigationBarTitleText: 'page title'
})
```
#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| navigationBarTitleText  | 当前页面标题 | 否 | string | - |
