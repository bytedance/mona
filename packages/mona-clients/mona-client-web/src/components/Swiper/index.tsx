// TODO: rewrite this component
import React, { useEffect, useRef } from 'react';
import { SwiperProps } from '@bytedance/mona';
import cs from 'classnames';
import styles from './index.module.less';
import { useHandlers } from '../hooks';
import { useSlide, useTouch } from './hooks';

function findIndexFromItemId(children: React.ReactNode, itemId: string) {
  let resultIndex = 0;
  React.Children.forEach(children, (child, index) => {
    if (React.isValidElement(child) && child.props.itemId === itemId) {
      resultIndex = index;
    }
  });
  return resultIndex;
}

const Swiper: React.FC<SwiperProps> = props => {
  const {
    children,
    indicatorDots = false,
    indicatorColor = 'rgba(0, 0, 0, .3)',
    indicatorActiveColor = 'rgba(0, 0, 0, 1)',
    autoplay = false,
    current = 0,
    currentItemId = '',
    interval = 5000,
    previousMargin = '',
    nextMargin = '',
    displayMultipleItems = 1,
    duration = 500,
    circular = false,
    vertical = false,
    onChange,
    // not implemented
    onAnimationFinish,
    // not implemented
    onTransition,
    ...restProps
  } = props;
  const ref = useRef<HTMLDivElement | null>(null);
  const dotsRef = useRef<HTMLDivElement | null>(null);

  const { handleClassName, ...handleProps } = useHandlers(restProps);

  const total = React.Children.count(children);
  const currentIndex = currentItemId ? findIndexFromItemId(children, currentItemId) : current;

  const { next, prev, wrapperRef, activeIndex } = useSlide({
    current: currentIndex,
    total,
    vertical,
    duration,
    circular,
    onChange,
  });

  const touchParams = {
    left: vertical ? null : (e: React.TouchEvent) => next(false, e),
    right: vertical ? null : (e: React.TouchEvent) => prev(false, e),
    top: vertical ? (e: React.TouchEvent) => next(false, e) : null,
    bottom: vertical ? (e: React.TouchEvent) => prev(false, e) : null,
    duration,
  };
  const { handleTouchStart, handleTouchMove } = useTouch(touchParams);

  // autoplay
  useEffect(() => {
    if (!total) {
      return;
    }
    if (autoplay) {
      const timer = setInterval(() => {
        next(true);
      }, interval);

      return () => clearInterval(timer);
    }
    return () => {};
  }, [autoplay, interval, next, total]);

  if (!total) {
    return null;
  }

  return (
    <div ref={ref} {...handleProps} className={handleClassName(styles.swiper)}>
      <div
        ref={wrapperRef}
        onTouchStart={handleTouchStart}
        onTouchMove={handleTouchMove}
        className={cs(vertical ? styles['vertical-wrapper'] : styles.wrapper)}
        style={{ transitionDuration: '0ms', transform: 'translate3d(10px, 0px, 0px)' }}
      >
        {children}
      </div>
      {indicatorDots && (
        <div className={cs(styles.dots, { [styles.horizontal]: !vertical })} ref={dotsRef}>
          {(function () {
            const result = [];
            for (let i = 0; i < total; i++) {
              result.push(
                <span
                  className={styles.dot}
                  style={{ backgroundColor: activeIndex === i ? indicatorActiveColor : indicatorColor }}
                  key={i}
                ></span>
              );
            }
            return result;
          })()}
        </div>
      )}
    </div>
  );
};

export default Swiper;
