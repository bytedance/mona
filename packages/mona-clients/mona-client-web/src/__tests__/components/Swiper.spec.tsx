import React from 'react';
import { mount } from 'enzyme';
import Swiper from '../../components/Swiper';
import mountTest from '../../../../../tests/shared/mountTest';
import handlerTest from '../../../../../tests/shared/handlerTest';
import { SwiperProps } from '@bytedance/mona';

describe('web component: Swiper', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  })
  afterEach(() => {
    jest.useRealTimers();
  })
  mountTest(Swiper)
  handlerTest({ Component: Swiper, firstTag: 'div', props: { children: [<div key="1"></div>] } });

  it('should render correctly', () => {
    expect(mount(<Swiper />).render()).toMatchSnapshot();
  })
  
  it('should have right properties', () => {
    const props: SwiperProps = {
      indicatorDots: false,
      indicatorColor: 'red',
      indicatorActiveColor: 'blue',
      autoplay: true,
      current: 10,
      currentItemId: 'hello',
      interval: 10,
      previousMargin: '10px',
      nextMargin: '10px',
      displayMultipleItems: 2,
      duration: 10,
      circular: true,
      vertical: true,
      className: 'class-name',
      style: { color: 'red' },
    }
    const wrapper = mount(<Swiper {...props}>hello</Swiper>);
    expect(wrapper.render()).toMatchSnapshot();
  })
})
