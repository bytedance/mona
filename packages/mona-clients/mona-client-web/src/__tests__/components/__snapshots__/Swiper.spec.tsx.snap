// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`web component: Swiper should have right properties 1`] = `
LoadedCheerio {
  "0": <div
    class="swiper class-name  "
    style="color: red;"
  >
    <div
      class="vertical-wrapper"
      style="transition-duration: 0ms; transform: translate3d(0px, 0px, 0px;"
    >
      hello
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`web component: Swiper should render correctly 1`] = `
LoadedCheerio {
  "0": Document {
    "children": Array [],
    "endIndex": null,
    "next": null,
    "parent": null,
    "prev": null,
    "startIndex": null,
    "type": "root",
  },
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
